Mathlib/Order/KonigLemma.lean

```diff
@@ -0,0 +1,147 @@
+
+-/
+
+open Set
+section Sequence
+
+variable {α : Type*} [PartialOrder α] [IsStronglyAtomic α] {a b : α}
+
+/-- **Kőnig's infinity lemma** : if each element in a strongly atomic order
+is covered by only finitely many others, and `b` is an element with infinitely many things above it,
+then there is a sequence starting with `b` in which each element is covered by the next. -/
+theorem exists_seq_covby_of_forall_covby_finite (hfin : ∀ (a : α), {x | a ⋖ x}.Finite)
+    (hb : (Ici b).Infinite) : ∃ f : ℕ → α, f 0 = b ∧ ∀ i, f i ⋖ f (i+1) :=
+  let h := fun a : {a : α // (Ici a).Infinite} ↦
+    exists_covby_infinite_Ici_of_infinite_Ici a.2 (hfin a)
+  let ks : ℕ → {a : α // (Ici a).Infinite} := Nat.rec ⟨b, hb⟩ fun _ a ↦ ⟨_, (h a).choose_spec.2⟩
+  ⟨fun i ↦ (ks i).1, by simp [ks], fun i ↦ by simpa using (h (ks i)).choose_spec.1⟩
+
+/-- The sequence given by Kőnig's lemma as an order embedding -/
+theorem exists_orderEmbedding_covby_of_forall_covby_finite (hfin : ∀ (a : α), {x | a ⋖ x}.Finite)
+    (hb : (Ici b).Infinite) : ∃ f : ℕ ↪o α, f 0 = b ∧ ∀ i, f i ⋖ f (i+1) := by
+  obtain ⟨f, hf⟩ := exists_seq_covby_of_forall_covby_finite hfin hb
+  exact ⟨OrderEmbedding.ofStrictMono f (strictMono_nat_of_lt_succ (fun i ↦ (hf.2 i).lt)), hf⟩
+
+/-- A version of Kőnig's lemma where the sequence starts at the minimum of an infinite order.  -/
+theorem exists_orderEmbedding_covby_of_forall_covby_finite_of_bot [OrderBot α] [Infinite α]
+    (hfin : ∀ (a : α), {x | a ⋖ x}.Finite) : ∃ f : ℕ ↪o α, f 0 = ⊥ ∧ ∀ i, f i ⋖ f (i+1) :=
+  exists_orderEmbedding_covby_of_forall_covby_finite hfin (by simpa using infinite_univ)
+
+theorem GradeMinOrder.exists_nat_orderEmbedding_of_forall_covby_finite
+    [GradeMinOrder ℕ α] [OrderBot α] [Infinite α] (hfin : ∀ (a : α), {x | a ⋖ x}.Finite) :
+    ∃ f : ℕ ↪o α, f 0 = ⊥ ∧ (∀ i, f i ⋖ f (i+1)) ∧ ∀ i, grade ℕ (f i) = i := by
+  obtain ⟨f, h0, hf⟩ := exists_orderEmbedding_covby_of_forall_covby_finite_of_bot hfin
+  refine ⟨f, h0, hf, fun i ↦ ?_⟩
+  induction' i with i ih; simp [h0]
+  simpa [Nat.covBy_iff_succ_eq, ih, eq_comm] using CovBy.grade ℕ <| hf i
+
+end Sequence
+
+section Graded
+
+/-- A formulation of Kőnig's infinity lemma, useful in applications.
+Given a sequence `α 0, α 1, ...` of nonempty types with `α 0` a subsingleton,
+and a well-behaved family of projections `π : α j → α i` for all `i ≤ j`,
+if each term in each `α i` is the projection of only finitely many terms in `α (i+1)`,
+then we can find a sequence `(f 0 : α 0), (f 1 : α 1), ...`
+where `f i` is the projection of `f j` for all `i ≤ j`.
+
+In a typical application, the `α i` are function types with increasingly large domains,
+and `π hij (f : α j)` is the restriction of the domain of `f` to that of `α i`.
+In this case, the sequence given by the lemma is essentially a function whose domain
+is the limit of the `α i`. -/
+theorem exists_seq_forall_proj_of_forall_finite {α : ℕ → Type*} [Subsingleton (α 0)]
```

Comment:
Why is the Subsingleton assumption necessary? Maybe it's easier to bootstrap the theorem by having `α 0` be a singleton, but the full theorem shouldn't need this since you can restrict the inverse system to the preimages of that element.
```suggestion
theorem exists_seq_forall_proj_of_forall_finite {α : ℕ → Type*}
```

Reply:
It's false if the subsingleton assumption is removed; for instance, take `α i = ↑(Ici i : Set ℕ)` where `π hij ⟨x, _⟩ = ⟨x, _⟩`. 

It is true with the weaker assumption `Finite (a 0)`, but version seems a little artificial to me, since it just amounts to cutting off the root of the tree. Should I still generalize it that way?

Reply:
Ah right, I forgot that this Ici example applies here, even if you relax finiteness just at the root.

I guess `Subsingleton` is fine, but it would be nice to have the finiteness condition be more uniform. One way to think about it is that there's a singleton "`α (-1)`" that augments the system.

Reply:
Yes - that's the right intuition but awkward to formalize. I'll mark this as resolved, if that's good with you.

Reply:
Let's please make it more general. If you change your `theorem exists_seq_forall_proj_of_forall_finite` to `private lemma exists_seq_forall_proj_of_forall_finite_aux` then here's 
```lean
theorem exists_seq_forall_proj_of_forall_finite {α : ℕ → Type*} [Finite (α 0)]
    [∀ i, Nonempty (α i)] (π : {i j : ℕ} → (hij : i ≤ j) → α j → α i)
    (π_refl : ∀ ⦃i⦄ (a : α i), π rfl.le a = a)
    (π_trans : ∀ ⦃i j k⦄ (hij : i ≤ j) (hjk : j ≤ k) a, π hij (π hjk a) = π (hij.trans hjk) a)
    (hfin : ∀ i a, {b : α (i+1) | π (Nat.le_add_right i 1) b = a}.Finite) :
    ∃ f : (i : ℕ) → α i, ∀ ⦃i j⦄ (hij : i ≤ j), π hij (f j) = f i := by
  let α' : ℕ → Type _
    | 0 => PUnit
    | i + 1 => α i
  let π' : {i j : ℕ} → (h : i ≤ j) → α' j → α' i
    | 0, 0, _, x => x
    | 0, _, _, _ => PUnit.unit
    | i + 1, j + 1, h, x => π (Nat.pred_le_pred h) x
  have (i : ℕ) : Nonempty (α' i) := by cases i <;> infer_instance
  have ⟨f', h'⟩ := exists_seq_forall_proj_of_forall_finite_aux π' ?rfl ?trans ?fin
  · use fun i => f' (i + 1)
    intros i j h
    exact h' (Nat.succ_le_succ h)
  case rfl =>
    rintro (_ | i)
    · simp
    · apply π_refl
  case trans =>
    rintro (_ | i) (_ | j) (_ | k) <;> (try simp; done)
    simpa using fun hij => π_trans hij
  case fin =>
    rintro (_ | i)
    · simp [α', finite_univ]
    · apply hfin
```

There's probably also a way to modify the original proof to make use of `WithBot ((i : ℕ) × α i)`, but there's no need for going down that path (and I didn't manage to do it myself).

Reply:
I found a cleaner way with no auxiliary lemma. (Just pick a base element with infinitely many things above it, then apply the version of Konig with that element at the bottom).

Reply:
That's great! It's nice that it didn't need much of a change to the proof.

(In my formalization of Konig's lemma, the first step was to restrict to the "very extendable" elements in every set — those elements with infinitely many things above them — and then show that each such element is the image of such an element. Then it's an application of choice. Seems like a similar principle here, but just for the base case?)